plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted = false;
+            if (s3ObjectMetadata == null ||
+                    s3DirectoryMetadata == null ||
+                    MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
+                deleted = deletePrefix(path);
+            }
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
```

Comment:
Shouldn't this be `new Path(stripDirectorySuffix(path.toString()) + PATH_SEPARATOR);`? Seems like the intention for a case like: `s3://bucket/directory_$folder$` would be to recursively delete `s3://bucket/directory/` (note that the trailing slash is significant, we don't want to delete objects like `s3://bucket/directory_sibling.txt`)

Reply:
See that in `deletePrefix(Path)` method is called `listObjects(prefix, OptionalInt.empty(), true);` to which is added `PATH_SEPARATOR`.

```
        String key = keyFromPath(path);
        if (!key.isEmpty()) {
            key += PATH_SEPARATOR;
        }
```